From 5d1213178455a53d1c4fdd08477ea6859dd8a31d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ball=C3=B3=20Gy=C3=B6rgy?= <ballogyor@gmail.com>
Date: Mon, 24 Aug 2026 11:05:16 +0200
Subject: [PATCH] Add option to hide on close

Instead of hiding the main window from the taskbar, hide the window on
close if the option is checked. This works also on Wayland.
---
 src/iptux/DataSettings.cpp |  2 +-
 src/iptux/MainWindow.cpp   | 25 ++++++++++++-------------
 2 files changed, 13 insertions(+), 14 deletions(-)

diff --git a/src/iptux/DataSettings.cpp b/src/iptux/DataSettings.cpp
index 8f02b3b0..bb0ef63f 100644
--- a/src/iptux/DataSettings.cpp
+++ b/src/iptux/DataSettings.cpp
@@ -452,7 +452,7 @@ GtkWidget* DataSettings::CreateSystem() {
 #if HAVE_APPINDICATOR
   row++;
   widget = gtk_check_button_new_with_label(
-      _("Hide the taskbar when the main window is minimized"));
+      _("Hide the main window instead of quit on close"));
   gtk_grid_attach(GTK_GRID(box), widget, 0, row, 2, 1);
   g_datalist_set_data(&widset, "taskbar-check-widget", widget);
 #endif
diff --git a/src/iptux/MainWindow.cpp b/src/iptux/MainWindow.cpp
index a770162a..39f00c54 100644
--- a/src/iptux/MainWindow.cpp
+++ b/src/iptux/MainWindow.cpp
@@ -72,21 +72,22 @@ static const char* config_names[] = {
     [CFG_INFO_STYLE] = "mwin_info_style",
 };
 
-static void main_window_on_state_event(GtkWidget* self,
-                                       GdkEvent* event,
-                                       gpointer user_data) {
-  GdkWindowState event_state = event->window_state.new_window_state;
+static gboolean main_window_on_delete_event(GtkWidget* self,
+                                            GdkEvent *event,
+                                            gpointer user_data) {
   MainWindow* mwin = (MainWindow*)user_data;
   auto progdt = mwin->getApp()->getProgramData();
 
+#if HAVE_APPINDICATOR
   if (!progdt->isHideTaskbarWhenMainWindowIconified())
-    return;
+    return FALSE;
 
-  if (event_state & GDK_WINDOW_STATE_ICONIFIED) {
-    gtk_window_set_skip_taskbar_hint(GTK_WINDOW(self), TRUE);
-  } else {
-    gtk_window_set_skip_taskbar_hint(GTK_WINDOW(self), FALSE);
-  }
+  gtk_widget_hide(self);
+
+  return TRUE;
+#else
+  return FALSE;
+#endif
 }
 
 /**
@@ -535,9 +536,7 @@ GtkWidget* MainWindow::CreateMainWindow() {
   g_signal_connect(window, "configure-event",
                    G_CALLBACK(WindowConfig::on_configure_event), &windowConfig);
   g_signal_connect(window, "delete-event",
-                   G_CALLBACK(gtk_window_iconify_on_delete), nullptr);
-  g_signal_connect(window, "window-state-event",
-                   G_CALLBACK(main_window_on_state_event), this);
+                   G_CALLBACK(main_window_on_delete_event), this);
   return window;
 }
 
